src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -452,7 +454,7 @@ private boolean tryToLaunchNameNode(SchedulerDriver driver, Offer offer) {
     List<String> deadNameNodes = persistentState.getDeadNameNodes();
 
     if (deadNameNodes.isEmpty()) {
-      if (liveState.getNameNodeSize() == HDFSConstants.TOTAL_NAME_NODES) {
+      if (persistentState.getNameNodes().size() == HDFSConstants.TOTAL_NAME_NODES) {
```

Comment:
I think LiveState should just be a (write-through) cache of the PersistentState. That way the scheduler only interacts with one state abstraction, which then decides when to write to zk (on any state change) or when to read from zk (on startup/recovery).